runtime/src/mm/cpp/ThreadStateTest.cpp

```diff
@@ -13,39 +13,65 @@
 
 using namespace kotlin;
 
-TEST(ThreadStateTest, ThreadStateSwitch) {
+TEST(ThreadStateTest, StateSwitch) {
     std::thread t([]() {
         mm::ThreadRegistry::Instance().RegisterCurrentThread();
 
         auto* threadData = mm::ThreadRegistry::Instance().CurrentThreadData();
         auto initialState = threadData->state();
         EXPECT_EQ(mm::ThreadState::kRunnable, initialState);
 
-        mm::ThreadState oldState = mm::SwitchThreadState(mm::kNative);
+        mm::ThreadState oldState = mm::SwitchThreadState(threadData, mm::ThreadState::kNative);
         EXPECT_EQ(initialState, oldState);
-        EXPECT_EQ(mm::kNative, threadData->state());
+        EXPECT_EQ(mm::ThreadState::kNative, threadData->state());
 
         // Check functions exported for the compiler too.
         Kotlin_mm_switchThreadStateRunnable();
-        EXPECT_EQ(mm::kRunnable, threadData->state());
+        EXPECT_EQ(mm::ThreadState::kRunnable, threadData->state());
 
         Kotlin_mm_switchThreadStateNative();
-        EXPECT_EQ(mm::kNative, threadData->state());
+        EXPECT_EQ(mm::ThreadState::kNative, threadData->state());
     });
     t.join();
 }
 
-TEST(ThreadStateTest, ThreadStateGuard) {
+TEST(ThreadStateTest, StateGuard) {
     std::thread t([]() {
         mm::ThreadRegistry::Instance().RegisterCurrentThread();
         auto* threadData = mm::ThreadRegistry::Instance().CurrentThreadData();
         auto initialState = threadData->state();
         EXPECT_EQ(mm::ThreadState::kRunnable, initialState);
         {
-            mm::ThreadStateGuard guard(mm::ThreadState::kNative);
+            mm::ThreadStateGuard guard(threadData, mm::ThreadState::kNative);
             EXPECT_EQ(mm::ThreadState::kNative, threadData->state());
         }
         EXPECT_EQ(initialState, threadData->state());
     });
     t.join();
 }
+
+TEST(ThreadStateDeathTest, StateAsserts) {
+    testing::FLAGS_gtest_death_test_style="threadsafe";
```

Comment:
Why is this needed?

Reply:
To execute a death test, GTest forks the test process. Since forking in a multithreaded environment may cause problems, GTest [supports](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#how-it-works) two death test styles: `fast` and `threadsafe`:
 - In the fast mode, the tested statement is executed right after the fork. If there are more than one thread at the moment of the fork (like in our case), GTest shows a corresponding warning.
 - In the thread safe mode, GTest restarts the test binary with additional flags to run only this particular death test. 

These particular tests seem to be safe for the `fast` mode but I still enabled the `threadsafe` mode to avoid GTest's warnings. The thread safe mode is 3 times slower for these tests but it's still a matter of tens of milliseconds on my machine (13 ms in the fast mode against 40 ms in the threadsafe mode).

There are also alternatives:

1. Move registering the thread to the `EXPECT_DEATH` and get rid of threads at all:

```C++
EXPECT_DEATH({
        // EXPECT_DEATH forks the test process so it's ok to register the main thread.
        auto* threadData = mm::ThreadRegistry::Instance().RegisterCurrentThread()->Get();
        mm::AssertThreadState(threadData, mm::ThreadState::kNative);
    }, "runtime assert: Unexpected thread state. Expected: NATIVE. Actual: RUNNABLE");
```

2. Create a `ThreadData` manually instead of calling `RegisterCurrentThread`. Not sure if it's a good idea though.

What is your opinion? Which alternative is better?

Reply:
I see. Let's make `threadsafe` a default by putting it into `TestLauncher.cpp`?